libvmaf/src/predict.c

```diff
@@ -61,30 +61,143 @@ static int denormalize(const VmafModel *model, double *prediction)
+        *alpha = 1.0;  // both points are the same
+        *beta = 0.0;
+    }
+    else if (p1.x == 0) {
+        *beta = p1.y;
+        *alpha = (p2.y - *beta) / p2.x;
+    }
+    else {
+        *beta = (p2.y * (p1.y - p1.x)) / (p2.x - p1.x);
+        *alpha = (p1.y - *beta) / p1.x;
+    }
+
+    return 0;
+}
+
+int piecewise_linear_mapping(double x, Point *knots, unsigned n_knots, double *y) {
+
+    double slope, offset;
+
+    if (n_knots <= 1)
+        return EINVAL;
+    unsigned n_seg = n_knots - 1;
+
+    *y = 0.0;
+
+    // construct the function
+    for (unsigned idx = 0; idx < n_seg; idx++) {
+        if (!(knots[idx].x <  knots[idx + 1].x &&
+              knots[idx].y <= knots[idx + 1].y))
+            return EINVAL;
+
+        bool cond0 = knots[idx].x <= x;
+        bool cond1 = x <= knots[idx + 1].x;
+
+        if (knots[idx].y == knots[idx + 1].y) {  // the segment is horizontal
+            if (cond0 && cond1)
+                *y = knots[idx].y;
+
+            if (idx == 0) {
+                // for points below the defined range
+                if (x < knots[idx].x)
+                    *y = knots[idx].y;
+            }
+            else if (idx == n_seg - 1) {
+                // for points above the defined range
+                if (x > knots[idx + 1].x)
+                    *y = knots[idx].y;
+            }
+        } else {
+            find_linear_function_parameters(knots[idx], knots[idx + 1],
+                                            &slope, &offset);
+
+            if (cond0 && cond1)
+                *y = slope * x + offset;
+
+            if (idx == 0) {
+                // for points below the defined range
+                if (x < knots[idx].x)
+                    *y = slope * x + offset;
+            }
+            else if (idx == n_seg - 1) {
+                // for points above the defined range
+                if (x > knots[idx + 1].x)
+                    *y = slope * x + offset;
+            }
+        }
+    }
+
+    return 0;
+}
+
+/*  Reproducing the logic in quality_runner.VmafQualityRunner.transform_score().
+    Transform final quality score in the following optional steps (in this
+    order):
+    1) polynomial mapping. e.g. {'p0': 1, 'p1': 1, 'p2': 0.5} means
+    transform through 1 + x + 0.5 * x^2. For now, only support polynomail
+    up to 2nd-order.
+    2) piecewise-linear mapping, where the change points are defined in
+    'knots', in the form of [[x0, y0], [x1, y1], ...].
+    3) rectification, supporting 'out_lte_in' (output is less than or equal
+    to input) and 'out_gte_in' (output is greater than or equal to input).
+ */
+static int transform(const VmafModel *model, double *y_in,
                      enum VmafModelFlags flags)
 {
     if (!model->score_transform.enabled)
         return 0;
     if (flags & VMAF_MODEL_FLAG_DISABLE_TRANSFORM)
         return 0;
 
-    double value = 0.;
-    double p = *prediction;
+    double y_stage, y_out;
+
+    // polynomial mapping
```

Comment:
Might be better to have a `VmafTransformMappingType` and use a switch here. Set the default (0) to be `VMAF_TRANSFORM_MAPPING_TYPE_POLYNOMIAL` for backwards compatibility.

Reply:
It is doing cascading of several transforms instead of just one. Probably OK to keep it this way for now (as it is in sync with Python code), and if it grows substantially in the future, we can always refactor to make it better.